api/src/main/kotlin/io/spine/protodata/renderer/SourceFile.kt

```diff
@@ -283,9 +328,110 @@ internal constructor(
     }
 }
 
+/**
+ * A fluent builder for inserting code into pre-prepared inline insertion points.
+ *
+ * @see SourceFile.atInline
+ */
+public sealed interface SourceAtPointInline {
+    /**
+     * Adds the specified code fragment into the insertion point.
+     *
+     * When called multiple times for the same insertion point, the code that is added last
+     * will appear first in the file, since new code fragments are always added right after
+     * the insertion point regardless of if it's been used before.
+     */
+    public fun add(codeFragment: String)
+}
+
+/**
+ * A [SourceAtPointInline] at a point that could not be located.
+ *
+ * No code is generated by this implementation.
+ */
+internal object Unlocated : SourceAtPointInline, Logging {
+    override fun add(codeFragment: String) {
+        // Do nothind.
+    }
+}
+
+/**
+ * A [SourceAtPointInline] at a point at a specific point or points in the file.
+ *
+ * The points are defined by the insertion point specified via the [point].
+ */
+internal class Located(
+    private val file: SourceFile,
+    private val point: InsertedPoint
+) : SourceAtPointInline {
+
+    override fun add(codeFragment: String) {
+        checkNoSeparator(codeFragment)
+        val sourceLines = file.lines()
+        val updatedLines = ArrayList(sourceLines)
+        sourceLines.asSequence()
+            .mapIndexed { index, line -> CodeLine(index, line) }
+            .map { line -> line.insertInline(point, codeFragment) }
+            .forEach { (index, line) -> updatedLines[index] = line }
+        file.updateLines(updatedLines)
+    }
+}
+
+private fun String.splitByIndexes(indexes: List<Int>): List<String> = buildList {
+    val idxs = buildList(indexes.size + 2) {
+        add(0)
+        addAll(indexes)
+        add(length)
+    }
+    idxs.forEachIndexed { listIndex, stringIndex ->
+        if (listIndex < idxs.size - 1) {
+            val nextIndex = idxs[listIndex + 1]
+            add(substring(stringIndex, nextIndex))
+        }
+    }
+}
+
+private fun String.findInsertionIndexes(point: InsertedPoint): List<Int> {
+    val comment = point.representationInCode
+    var index = 0
+    return buildList {
+        while (true) {
+            val rawIndex = indexOf(comment, startIndex = index)
+            if (rawIndex >= 0) {
+                index = rawIndex + comment.length
+                add(index)
+            } else {
+                break
+            }
+        }
+    }
+}
+
 private fun Iterable<String>.linesToCode(indentLevel: Int): String =
     joinToString(lineSeparator()) {
         INDENT.repeat(indentLevel) + it
     }
 
 private const val INDENT: String = "    "
+
+/**
+ * A numbered line of code in a file.
+ */
+private data class CodeLine(val lineIndex: Int, val content: String) {
```

Comment:
Ha, see, we do need numbers together with lines!  Why isn't it a `Line` in a Text library then?